server/src/main/java/org/elasticsearch/common/collect/ImmutableOpenMap.java

```diff
@@ -136,6 +138,24 @@ public void remove() {
         };
     }
 
+    /**
+     * Returns a {@link Set} view of the keys contained in this map.
+     */
+    public Set<KType> keySet() {
+        final Iterator<KType> keysIterator = keysIt();
+        return new AbstractSet<>() {
```

Comment:
This doesn't look like it provides a complete set view of the keys, there should also be some other methods overriden for contains/etc?

Reply:
`AbstractSet` extends `AbstractCollection` already provides skeleton implementations for `contains`, `remove`, `toArray`, etc based on the `iterator` method.

Reply:
I think overriding `contains` makes sense since it will be very inefficiently implemented otherwise.

Reply:
I think `add` and 'addAll',  should also be overriden to throw `OperationNotSupportedException`. `remove` and `removeAll` as well, but the behavior can be debated. I think we should make this operate just as `Map.keySet()`, which means the returned set should be "live", so removing a key removes the underlying key from the map. The reason I think we should mimic `Map.keySet()` is so that we can work towards actually making `ImmutableOpenMap implement Map`. This will allow consumers to operate on it as a Map. It is a long term idea, but this would be a start.

Reply:
I think we should leave those methods throwing an `UnsupportedOperationException`, since the `ImmutableOpenMap` is immutable. We could overload them to refine the message but I find it unnecessary, since it is a bug/coding error anyway.

Reply:
That’s a good point, and why I thought overriding add/addAll was necessary. It hadn’t occurred to me that AbstractMap threw UOE for these already, of course it must.